MoreLinq.Test/CountDownTest.cs

```diff
@@ -0,0 +1,56 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 Atif Aziz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq.Test
+{
+    using System.Collections;
+    using System.Collections.Generic;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class CountDownTest
+    {
+        [Test]
+        public void IsLazy()
+        {
+            new BreakingSequence<object>()
+                .CountDown(42, BreakingFunc.Of<object, int?, object>());
+        }
+
+        static readonly IEnumerable<TestCaseData> Data =
+            from e in new[]
+            {
+                new { Count = -1, CountDown = new int?[] { null, null, null, null, null } },
+                new { Count =  0, CountDown = new int?[] { null, null, null, null, null } },
+                new { Count =  1, CountDown = new int?[] { null, null, null, null,    0 } },
+                new { Count =  2, CountDown = new int?[] { null, null, null,    1,    0 } },
+                new { Count =  3, CountDown = new int?[] { null, null,    2,    1,    0 } },
+                new { Count =  4, CountDown = new int?[] { null,    3,    2,    1,    0 } },
+                new { Count =  5, CountDown = new int?[] {    4,    3,    2,    1,    0 } },
+                new { Count =  6, CountDown = new int?[] {    4,    3,    2,    1,    0 } },
+                new { Count =  7, CountDown = new int?[] {    4,    3,    2,    1,    0 } },
+            }
+            let xs = Enumerable.Range(1, 5)
+            select new TestCaseData(xs, e.Count)
+                .Returns(xs.Zip(e.CountDown, (x, cd) => new { X = x, Countdown = cd }))
+                .SetName($"{nameof(CountDown)}([{xs.First()}..{xs.Last()}], {e.Count})");
+
+        [TestCaseSource(nameof(Data))]
+        public IEnumerable CountDown(IEnumerable<int> xs, int count) =>
+            xs.CountDown(count, (x, cd) => new { X = x, Countdown = cd });
```

Comment:
The return type is not strictly important. It could even be `object` but I can change to tuples if it makes things clearer. One other reason for preferring an anonymous type is that they have naturally good formatting, so when a test fails, instead of seeing something like expected `(1, -1)`, as you would for a tuple, you see `{ X = 1, Countdown = -1 }`.